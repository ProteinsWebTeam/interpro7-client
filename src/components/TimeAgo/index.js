// @flow
import React, { PureComponent } from 'react';
import T from 'prop-types';
import { format } from 'timeago.js';
import { sleep, schedule } from 'timing-functions';
import { formatShortDate, formatLongDateTime } from 'utils/date';

import random from 'utils/random';

const ONE_MINUTE = 60000;

/*:: type Props = {
  date: Date,
  title?: string,
  noTitle?: boolean,
  noUpdate?: boolean,
};*/

class TimeAgo extends PureComponent /*:: <Props> */ {
  /* ::
    _ref: { current: HTMLElement | null };
    __delay: number;
  */
  static propTypes = {
    date: T.instanceOf(Date).isRequired,
    title: T.string,
    noTitle: T.bool,
    noUpdate: T.bool,
  };

  constructor(props /*: Props */) {
    super(props);
    this._ref = React.createRef();
  }

  async componentDidMount() {
    if (this.props.noUpdate) return;
    await sleep(this._delay);
    await schedule();
    // infinite loop while mounted
    while (this._ref.current) {
      // $FlowIgnore
      this._ref.current.textContent = format(this.props.date);
      await sleep(this._delay);
      await schedule();
    }
  }

  // delay before re-rendering will slowly grow everytime by up to 1 minute
  get _delay() {
    this.__delay = (this.__delay || 0) + random(0, ONE_MINUTE);
    return this.__delay;
  }

  render() {
    const { date, title, noTitle } = this.props;
    let _title;
    if (!noTitle) {
      _title = title || formatShortDate(date);
    }
    return (
      <time dateTime={formatLongDateTime(date)} title={_title} ref={this._ref}>
        {format(date)}
      </time>
    );
  }
}

export default TimeAgo;
