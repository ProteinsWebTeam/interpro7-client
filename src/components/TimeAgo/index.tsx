import React, { PureComponent, RefObject } from 'react';
import { format } from 'timeago.js';
import { sleep, schedule } from 'timing-functions';
import { formatShortDate, formatLongDateTime } from 'utils/date';

import random from 'utils/random';

const ONE_MINUTE = 60000;

type Props = {
  date: Date;
  title?: string;
  noTitle?: boolean;
  noUpdate?: boolean;
};

class TimeAgo extends PureComponent<Props> {
  _ref: RefObject<HTMLTimeElement>;
  #delay = 0;

  constructor(props: Props) {
    super(props);
    this._ref = React.createRef();
  }

  async componentDidMount() {
    if (this.props.noUpdate) return;
    await sleep(this._delay);
    await schedule();
    // infinite loop while mounted
    while (this._ref.current) {
      // $FlowIgnore
      this._ref.current.textContent = format(this.props.date);
      await sleep(this._delay);
      await schedule();
    }
  }

  // delay before re-rendering will slowly grow everytime by up to 1 minute
  get _delay() {
    this.#delay = (this.#delay || 0) + random(0, ONE_MINUTE);
    return this.#delay;
  }

  render() {
    const { date, title, noTitle } = this.props;
    let _title;
    if (!noTitle) {
      _title = title || formatShortDate(date);
    }
    return (
      <time dateTime={formatLongDateTime(date)} title={_title} ref={this._ref}>
        {format(date)}
      </time>
    );
  }
}

export default TimeAgo;
