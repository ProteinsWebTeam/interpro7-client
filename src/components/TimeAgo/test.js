// @flow
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import TimeAgo from '.';

const renderer = new ShallowRenderer();

describe('<TimeAgo />', () => {
  test('should render', () => {
    renderer.render(
      <TimeAgo
        date={new Date('June 17, 1998 01:00:00')}
        noUpdate={true}
        title={'17/06/1998'}
      />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
