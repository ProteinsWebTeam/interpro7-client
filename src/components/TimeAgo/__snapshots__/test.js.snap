// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeAgo /> should render 1`] = `
<time
  dateTime="17-06-1998 01:00:00"
  title="17/06/1998"
>
  24 years ago
</time>
`;
